// @nx/angular/src/generators/directive/directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.directiveGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const utils_1 = require("../utils");
const lib_1 = require("./lib");
function directiveGenerator(tree, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        (0, lib_1.validateOptions)(tree, schema);
        const options = (0, lib_1.normalizeOptions)(tree, schema);
        const directiveNames = (0, devkit_1.names)(options.name);
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, 'files'), options.directory, {
            selector: options.selector,
            directiveClassName: directiveNames.className,
            directiveFileName: directiveNames.fileName,
            standalone: options.standalone,
            tpl: '',
        });
        if (options.skipTests) {
            const pathToSpecFile = (0, devkit_1.joinPathFragments)(options.directory, `${directiveNames.fileName}.directive.spec.ts`);
            tree.delete(pathToSpecFile);
        }
        if (!options.skipImport && !options.standalone) {
            const modulePath = (0, utils_1.findModule)(tree, options.path, options.module);
            (0, utils_1.addToNgModule)(tree, options.path, modulePath, directiveNames.fileName, `${directiveNames.className}Directive`, `${directiveNames.fileName}.directive`, 'declarations', options.flat, options.export);
        }
        if (!options.skipFormat) {
            yield (0, devkit_1.formatFiles)(tree);
        }
    });
}
exports.directiveGenerator = directiveGenerator;
exports.default = directiveGenerator;
