// @angular-material-components/color-picker/esm2020/lib/color-picker.module.mjs
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatLegacyButtonModule as MatButtonModule } from '@angular/material/legacy-button';
import { MatLegacyCardModule as MatCardModule } from '@angular/material/legacy-card';
import { MatLegacyDialogModule as MatDialogModule } from '@angular/material/legacy-dialog';
import { MatIconModule } from '@angular/material/icon';
import { MatLegacyInputModule as MatInputModule } from '@angular/material/legacy-input';
import { MatLegacyRadioModule as MatRadioModule } from '@angular/material/legacy-radio';
import { NgxMatColorCanvasComponent, NgxMatColorCollectionComponent, NgxMatColorPaletteComponent, NgxMatColorPickerComponent, NgxMatColorPickerInput, NgxMatColorSliderComponent, NgxMatColorToggleComponent, NGX_MAT_COLOR_PICKER_SCROLL_STRATEGY_FACTORY_PROVIDER, NgxMatColorPickerContentComponent } from './components';
import { NumericColorInputDirective } from './directives';
import { ColorAdapter } from './services';
import * as i0 from "@angular/core";
export class NgxMatColorPickerModule {
}
/** @nocollapse */ NgxMatColorPickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatColorPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ NgxMatColorPickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.2", ngImport: i0, type: NgxMatColorPickerModule, declarations: [NgxMatColorPaletteComponent,
        NgxMatColorCanvasComponent,
        NgxMatColorCollectionComponent,
        NgxMatColorSliderComponent,
        NumericColorInputDirective,
        NgxMatColorPickerContentComponent,
        NgxMatColorPickerComponent,
        NgxMatColorToggleComponent,
        NgxMatColorPickerInput], imports: [CommonModule,
        MatInputModule,
        MatButtonModule,
        MatCardModule,
        MatRadioModule,
        FormsModule,
        ReactiveFormsModule,
        MatDialogModule,
        PortalModule,
        MatIconModule], exports: [NgxMatColorToggleComponent,
        NgxMatColorPickerInput,
        NgxMatColorPickerComponent] });
/** @nocollapse */ NgxMatColorPickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatColorPickerModule, providers: [
        ColorAdapter,
        NGX_MAT_COLOR_PICKER_SCROLL_STRATEGY_FACTORY_PROVIDER
    ], imports: [CommonModule,
        MatInputModule,
        MatButtonModule,
        MatCardModule,
        MatRadioModule,
        FormsModule,
        ReactiveFormsModule,
        MatDialogModule,
        PortalModule,
        MatIconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatColorPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxMatColorPaletteComponent,
                        NgxMatColorCanvasComponent,
                        NgxMatColorCollectionComponent,
                        NgxMatColorSliderComponent,
                        NumericColorInputDirective,
                        NgxMatColorPickerContentComponent,
                        NgxMatColorPickerComponent,
                        NgxMatColorToggleComponent,
                        NgxMatColorPickerInput
                    ],
                    imports: [
                        CommonModule,
                        MatInputModule,
                        MatButtonModule,
                        MatCardModule,
                        MatRadioModule,
                        FormsModule,
                        ReactiveFormsModule,
                        MatDialogModule,
                        PortalModule,
                        MatIconModule
                    ],
                    exports: [
                        NgxMatColorToggleComponent,
                        NgxMatColorPickerInput,
                        NgxMatColorPickerComponent
                    ],
                    entryComponents: [
                        NgxMatColorPickerContentComponent
                    ],
                    providers: [
                        ColorAdapter,
                        NGX_MAT_COLOR_PICKER_SCROLL_STRATEGY_FACTORY_PROVIDER
                    ]
                }]
        }] });
