// @nx/angular/src/generators/setup-ssr/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const ast_utils_1 = require("../../../utils/nx-devkit/ast-utils");
function normalizeOptions(tree, options) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const isStandaloneApp = (0, ast_utils_1.isNgStandaloneApp)(tree, options.project);
    return {
        project: options.project,
        appId: (_a = options.appId) !== null && _a !== void 0 ? _a : 'serverApp',
        main: (_b = options.main) !== null && _b !== void 0 ? _b : 'main.server.ts',
        serverFileName: (_c = options.serverFileName) !== null && _c !== void 0 ? _c : 'server.ts',
        serverPort: (_d = options.serverPort) !== null && _d !== void 0 ? _d : 4000,
        rootModuleFileName: (_e = options.rootModuleFileName) !== null && _e !== void 0 ? _e : 'app.server.module.ts',
        rootModuleClassName: (_f = options.rootModuleClassName) !== null && _f !== void 0 ? _f : 'AppServerModule',
        skipFormat: (_g = options.skipFormat) !== null && _g !== void 0 ? _g : false,
        standalone: (_h = options.standalone) !== null && _h !== void 0 ? _h : isStandaloneApp,
    };
}
exports.normalizeOptions = normalizeOptions;
