// @nx/angular/src/executors/package/ng-packagr-adjustments/ng-package/entry-point/compile-ngc.transform.js
"use strict";
/**
 * Adapted from the original ng-packagr source.
 *
 * Changes made:
 * - Use our own StylesheetProcessor files instead of the ones provide by ng-packagr.
 * - Support ngcc for Angular < 16.
 * - Support ESM2020 for Angular < 16.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileNgcTransformFactory = void 0;
const tslib_1 = require("tslib");
const transform_1 = require("ng-packagr/lib/graph/transform");
const nodes_1 = require("ng-packagr/lib/ng-package/nodes");
const tsconfig_1 = require("ng-packagr/lib/ts/tsconfig");
const ora = require("ora");
const path = require("path");
const ts = require("typescript");
const angular_version_utils_1 = require("../../../../utilities/angular-version-utils");
const compile_source_files_1 = require("../../ngc/compile-source-files");
const ng_compiler_cli_1 = require("../../utils/ng-compiler-cli");
const compileNgcTransformFactory = (StylesheetProcessor, options) => {
    return (0, transform_1.transformFromPromise)((graph) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        var _a;
        var _b;
        const spinner = ora({
            hideCursor: false,
            discardStdin: false,
        });
        try {
            const entryPoint = graph.find((0, nodes_1.isEntryPointInProgress)());
            const entryPoints = graph.filter(nodes_1.isEntryPoint);
            // Add paths mappings for dependencies
            const tsConfig = (0, tsconfig_1.setDependenciesTsConfigPaths)(entryPoint.data.tsConfig, entryPoints);
            const angularVersion = (0, angular_version_utils_1.getInstalledAngularVersionInfo)();
            // Compile TypeScript sources
            const { declarations } = entryPoint.data.destinationFiles;
            const esmModulePath = angularVersion.major < 16
                ? entryPoint.data.destinationFiles.esm2020
                : entryPoint.data.destinationFiles.esm2022;
            const { basePath, cssUrl, styleIncludePaths } = entryPoint.data.entryPoint;
            const { moduleResolutionCache } = entryPoint.cache;
            spinner.start(`Compiling with Angular sources in Ivy ${tsConfig.options.compilationMode || 'full'} compilation mode.`);
            let ngccProcessor;
            if (angularVersion && angularVersion.major < 16) {
                ngccProcessor =
                    new (require('ng-packagr/lib/ngc/ngcc-processor').NgccProcessor)(yield (0, ng_compiler_cli_1.ngccCompilerCli)(), entryPoint.cache.ngccProcessingCache, tsConfig.project, tsConfig.options, entryPoints);
                if (!entryPoint.data.entryPoint.isSecondaryEntryPoint) {
                    // Only run the async version of NGCC during the primary entrypoint processing.
                    yield ngccProcessor.process();
                }
            }
            (_a = (_b = entryPoint.cache).stylesheetProcessor) !== null && _a !== void 0 ? _a : (_b.stylesheetProcessor = new StylesheetProcessor(basePath, cssUrl, styleIncludePaths, options.cacheEnabled && options.cacheDirectory, options.watch, options.tailwindConfig));
            yield (0, compile_source_files_1.compileSourceFiles)(graph, tsConfig, moduleResolutionCache, {
                outDir: path.dirname(esmModulePath),
                declarationDir: path.dirname(declarations),
                declaration: true,
                target: angularVersion.major >= 16
                    ? ts.ScriptTarget.ES2022
                    : ts.ScriptTarget.ES2020,
            }, entryPoint.cache.stylesheetProcessor, ngccProcessor, options.watch);
        }
        catch (error) {
            spinner.fail();
            throw error;
        }
        spinner.succeed();
        return graph;
    }));
};
exports.compileNgcTransformFactory = compileNgcTransformFactory;
