// @nx/js/src/utils/typescript/ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readTsConfigPaths = exports.addTsConfigPath = exports.getRootTsConfigFileName = exports.getRootTsConfigPath = exports.getRelativePathToRootTsConfig = exports.getRootTsConfigPathInTree = exports.readTsConfig = void 0;
const devkit_1 = require("@nx/devkit");
const fs_1 = require("fs");
const path_1 = require("path");
const ensure_typescript_1 = require("./ensure-typescript");
let tsModule;
function readTsConfig(tsConfigPath) {
    if (!tsModule) {
        tsModule = require('typescript');
    }
    const readResult = tsModule.readConfigFile(tsConfigPath, tsModule.sys.readFile);
    return tsModule.parseJsonConfigFileContent(readResult.config, tsModule.sys, (0, path_1.dirname)(tsConfigPath));
}
exports.readTsConfig = readTsConfig;
function getRootTsConfigPathInTree(tree) {
    for (const path of ['tsconfig.base.json', 'tsconfig.json']) {
        if (tree.exists(path)) {
            return path;
        }
    }
    return 'tsconfig.base.json';
}
exports.getRootTsConfigPathInTree = getRootTsConfigPathInTree;
function getRelativePathToRootTsConfig(tree, targetPath) {
    return (0, devkit_1.offsetFromRoot)(targetPath) + getRootTsConfigPathInTree(tree);
}
exports.getRelativePathToRootTsConfig = getRelativePathToRootTsConfig;
function getRootTsConfigPath() {
    const tsConfigFileName = getRootTsConfigFileName();
    return tsConfigFileName ? (0, path_1.join)(devkit_1.workspaceRoot, tsConfigFileName) : null;
}
exports.getRootTsConfigPath = getRootTsConfigPath;
function getRootTsConfigFileName(tree) {
    for (const tsConfigName of ['tsconfig.base.json', 'tsconfig.json']) {
        const pathExists = tree
            ? tree.exists(tsConfigName)
            : (0, fs_1.existsSync)((0, path_1.join)(devkit_1.workspaceRoot, tsConfigName));
        if (pathExists) {
            return tsConfigName;
        }
    }
    return null;
}
exports.getRootTsConfigFileName = getRootTsConfigFileName;
function addTsConfigPath(tree, importPath, lookupPaths) {
    (0, devkit_1.updateJson)(tree, getRootTsConfigPathInTree(tree), (json) => {
        var _a;
        const c = json.compilerOptions;
        (_a = c.paths) !== null && _a !== void 0 ? _a : (c.paths = {});
        if (c.paths[importPath]) {
            throw new Error(`You already have a library using the import path "${importPath}". Make sure to specify a unique one.`);
        }
        c.paths[importPath] = lookupPaths;
        return json;
    });
}
exports.addTsConfigPath = addTsConfigPath;
function readTsConfigPaths(tsConfig) {
    var _a;
    tsConfig !== null && tsConfig !== void 0 ? tsConfig : (tsConfig = getRootTsConfigPath());
    try {
        if (!tsModule) {
            tsModule = (0, ensure_typescript_1.ensureTypescript)();
        }
        let config;
        if (typeof tsConfig === 'string') {
            const configFile = tsModule.readConfigFile(tsConfig, tsModule.sys.readFile);
            config = tsModule.parseJsonConfigFileContent(configFile.config, tsModule.sys, (0, path_1.dirname)(tsConfig));
        }
        else {
            config = tsConfig;
        }
        if ((_a = config.options) === null || _a === void 0 ? void 0 : _a.paths) {
            return config.options.paths;
        }
        else {
            return null;
        }
    }
    catch (e) {
        return null;
    }
}
exports.readTsConfigPaths = readTsConfigPaths;
