// @nx/angular/src/generators/scam/lib/convert-component-to-scam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertComponentToScam = void 0;
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const ensure_typescript_1 = require("@nx/js/src/utils/typescript/ensure-typescript");
let tsModule;
function convertComponentToScam(tree, options) {
    var _a;
    if (!tree.exists(options.filePath)) {
        throw new Error(`Couldn't find component at path ${options.filePath} to add SCAM setup.`);
    }
    const componentNames = (0, devkit_1.names)(options.name);
    const typeNames = (0, devkit_1.names)((_a = options.type) !== null && _a !== void 0 ? _a : 'component');
    const componentClassName = `${componentNames.className}${typeNames.className}`;
    if (!tsModule) {
        tsModule = (0, ensure_typescript_1.ensureTypescript)();
    }
    if (options.inlineScam) {
        const currentComponentContents = tree.read(options.filePath, 'utf-8');
        let source = tsModule.createSourceFile(options.filePath, currentComponentContents, tsModule.ScriptTarget.Latest, true);
        source = (0, js_1.insertImport)(tree, source, options.filePath, 'NgModule', '@angular/core');
        source = (0, js_1.insertImport)(tree, source, options.filePath, 'CommonModule', '@angular/common');
        let updatedComponentSource = source.getText();
        updatedComponentSource = `${updatedComponentSource}${getNgModuleDeclaration(componentClassName)}`;
        tree.write(options.filePath, updatedComponentSource);
        return;
    }
    const moduleFilePath = (0, devkit_1.joinPathFragments)(options.directory, `${componentNames.fileName}.module.ts`);
    tree.write(moduleFilePath, getModuleFileContent(componentClassName, options.fileName));
}
exports.convertComponentToScam = convertComponentToScam;
function getModuleFileContent(componentClassName, componentFileName) {
    return `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ${componentClassName} } from './${componentFileName}';
${getNgModuleDeclaration(componentClassName)}`;
}
function getNgModuleDeclaration(componentClassName) {
    return `
@NgModule({
  imports: [CommonModule],
  declarations: [${componentClassName}],
  exports: [${componentClassName}],
})
export class ${componentClassName}Module {}`;
}
