// @nx/eslint-plugin/src/utils/project-graph-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readProjectGraph = exports.ensureGlobalProjectGraph = void 0;
const devkit_1 = require("@nx/devkit");
const runtime_lint_utils_1 = require("./runtime-lint-utils");
const chalk = require("chalk");
const find_project_for_path_1 = require("nx/src/project-graph/utils/find-project-for-path");
const file_utils_1 = require("nx/src/project-graph/file-utils");
const internal_1 = require("@nx/js/src/internal");
const nx_deps_cache_1 = require("nx/src/project-graph/nx-deps-cache");
function ensureGlobalProjectGraph(ruleName) {
    /**
     * Only reuse graph when running from terminal
     * Enforce every IDE change to get a fresh nxdeps.json
     */
    if (!global.projectGraph ||
        !global.projectRootMappings ||
        !global.projectFileMap ||
        !(0, runtime_lint_utils_1.isTerminalRun)()) {
        const nxJson = (0, file_utils_1.readNxJson)();
        global.workspaceLayout = nxJson.workspaceLayout;
        /**
         * Because there are a number of ways in which the rule can be invoked (executor vs ESLint CLI vs IDE Plugin),
         * the ProjectGraph may or may not exist by the time the lint rule is invoked for the first time.
         */
        try {
            const projectGraph = (0, devkit_1.readCachedProjectGraph)();
            global.projectGraph = projectGraph;
            global.projectRootMappings = (0, find_project_for_path_1.createProjectRootMappings)(projectGraph.nodes);
            global.projectFileMap = (0, nx_deps_cache_1.readProjectFileMapCache)().projectFileMap;
            global.targetProjectLocator = new internal_1.TargetProjectLocator(projectGraph.nodes, projectGraph.externalNodes);
        }
        catch (_a) {
            const WARNING_PREFIX = `${chalk.reset.keyword('orange')('warning')}`;
            const RULE_NAME_SUFFIX = `${chalk.reset.dim(`@nx/${ruleName}`)}`;
            process.stdout
                .write(`${WARNING_PREFIX} No cached ProjectGraph is available. The rule will be skipped.
          If you encounter this error as part of running standard \`nx\` commands then please open an issue on https://github.com/nrwl/nx
          ${RULE_NAME_SUFFIX}\n`);
        }
    }
}
exports.ensureGlobalProjectGraph = ensureGlobalProjectGraph;
function readProjectGraph(ruleName) {
    ensureGlobalProjectGraph(ruleName);
    return {
        projectGraph: global.projectGraph,
        projectFileMap: global.projectFileMap,
        projectRootMappings: global.projectRootMappings,
        targetProjectLocator: global.targetProjectLocator,
    };
}
exports.readProjectGraph = readProjectGraph;
