// @nx/angular/src/generators/application/lib/add-e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addE2e = void 0;
const tslib_1 = require("tslib");
const cypress_1 = require("@nx/cypress");
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../../utils/versions");
const remove_scaffolded_e2e_1 = require("./remove-scaffolded-e2e");
function addE2e(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        (0, remove_scaffolded_e2e_1.removeScaffoldedE2e)(tree, options, options.ngCliSchematicE2ERoot);
        if (options.e2eTestRunner === 'cypress') {
            // TODO: This can call `@nx/web:static-config` generator when ready
            addFileServerTarget(tree, options, 'serve-static');
            yield (0, cypress_1.cypressProjectGenerator)(tree, {
                name: options.e2eProjectName,
                directory: options.directory,
                project: options.name,
                linter: options.linter,
                standaloneConfig: options.standaloneConfig,
                skipPackageJson: options.skipPackageJson,
                skipFormat: true,
            });
        }
    });
}
exports.addE2e = addE2e;
function addFileServerTarget(tree, options, targetName) {
    (0, devkit_1.addDependenciesToPackageJson)(tree, {}, { '@nx/web': versions_1.nxVersion });
    const projectConfig = (0, devkit_1.readProjectConfiguration)(tree, options.name);
    projectConfig.targets[targetName] = {
        executor: '@nx/web:file-server',
        options: {
            buildTarget: `${options.name}:build`,
            port: options.port,
        },
    };
    (0, devkit_1.updateProjectConfiguration)(tree, options.name, projectConfig);
}
