// @nx/angular/src/generators/application/lib/add-unit-test-runner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addUnitTestRunner = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const jest_1 = require("@nx/jest");
const test_runners_1 = require("../../../utils/test-runners");
function addUnitTestRunner(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.unitTestRunner === test_runners_1.UnitTestRunner.Jest) {
            yield (0, jest_1.jestProjectGenerator)(host, {
                project: options.name,
                setupFile: 'angular',
                supportTsx: false,
                skipSerializers: false,
                skipPackageJson: options.skipPackageJson,
                skipFormat: true,
            });
            const setupFile = (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'src', 'test-setup.ts');
            if (options.strict && host.exists(setupFile)) {
                const contents = host.read(setupFile, 'utf-8');
                host.write(setupFile, `// @ts-expect-error https://thymikee.github.io/jest-preset-angular/docs/getting-started/test-environment
globalThis.ngJest = {
  testEnvironmentOptions: {
    errorOnUnknownElements: true,
    errorOnUnknownProperties: true,
  },
};
${contents}`);
            }
        }
    });
}
exports.addUnitTestRunner = addUnitTestRunner;
