// @nx/angular/src/generators/application/lib/enable-strict-type-checking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.enableStrictTypeChecking = void 0;
const devkit_1 = require("@nx/devkit");
function enableStrictTypeChecking(host, options) {
    // This matches the settings defined by the Angular CLI https://angular.io/guide/strict-mode
    const compilerOptions = {
        forceConsistentCasingInFileNames: true,
        strict: true,
        noImplicitOverride: true,
        noPropertyAccessFromIndexSignature: true,
        noImplicitReturns: true,
        noFallthroughCasesInSwitch: true,
    };
    const appTsConfigPath = `${options.appProjectRoot}/tsconfig.json`;
    if (host.exists(appTsConfigPath)) {
        (0, devkit_1.updateJson)(host, appTsConfigPath, (json) => {
            json.compilerOptions = Object.assign(Object.assign({}, json.compilerOptions), compilerOptions);
            json.angularCompilerOptions = {
                enableI18nLegacyMessageIdFormat: false,
                strictInjectionParameters: true,
                strictInputAccessModifiers: true,
                strictTemplates: true,
            };
            return json;
        });
    }
    const e2eTsConfigPath = `${options.e2eProjectRoot}/tsconfig.json`;
    if (host.exists(e2eTsConfigPath)) {
        (0, devkit_1.updateJson)(host, e2eTsConfigPath, (json) => {
            json.compilerOptions = Object.assign(Object.assign({}, json.compilerOptions), compilerOptions);
            return json;
        });
    }
}
exports.enableStrictTypeChecking = enableStrictTypeChecking;
