// @nx/angular/src/generators/utils/version-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versions = exports.addDependenciesToPackageJsonIfDontExist = exports.getInstalledPackageVersionInfo = exports.getInstalledPackageVersion = exports.getInstalledAngularVersionInfo = exports.getInstalledAngularMajorVersion = exports.getInstalledAngularVersion = void 0;
const devkit_1 = require("@nx/devkit");
const semver_1 = require("semver");
const latestVersions = require("../../utils/versions");
const versions_1 = require("../../utils/versions");
const backward_compatible_versions_1 = require("../../utils/backward-compatible-versions");
function getInstalledAngularVersion(tree) {
    var _a, _b;
    const pkgJson = (0, devkit_1.readJson)(tree, 'package.json');
    const installedAngularVersion = pkgJson.dependencies && pkgJson.dependencies['@angular/core'];
    if (!installedAngularVersion ||
        installedAngularVersion === 'latest' ||
        installedAngularVersion === 'next') {
        return (_a = (0, semver_1.clean)(versions_1.angularVersion)) !== null && _a !== void 0 ? _a : (0, semver_1.coerce)(versions_1.angularVersion).version;
    }
    return ((_b = (0, semver_1.clean)(installedAngularVersion)) !== null && _b !== void 0 ? _b : (0, semver_1.coerce)(installedAngularVersion).version);
}
exports.getInstalledAngularVersion = getInstalledAngularVersion;
function getInstalledAngularMajorVersion(tree) {
    return (0, semver_1.major)(getInstalledAngularVersion(tree));
}
exports.getInstalledAngularMajorVersion = getInstalledAngularMajorVersion;
function getInstalledAngularVersionInfo(tree) {
    const installedVersion = getInstalledAngularVersion(tree);
    return {
        version: installedVersion,
        major: (0, semver_1.major)(installedVersion),
    };
}
exports.getInstalledAngularVersionInfo = getInstalledAngularVersionInfo;
function getInstalledPackageVersion(tree, pkgName) {
    var _a;
    const { dependencies, devDependencies } = (0, devkit_1.readJson)(tree, 'package.json');
    const version = (_a = dependencies === null || dependencies === void 0 ? void 0 : dependencies[pkgName]) !== null && _a !== void 0 ? _a : devDependencies === null || devDependencies === void 0 ? void 0 : devDependencies[pkgName];
    return version;
}
exports.getInstalledPackageVersion = getInstalledPackageVersion;
function getInstalledPackageVersionInfo(tree, pkgName) {
    const version = getInstalledPackageVersion(tree, pkgName);
    return version ? { major: (0, semver_1.major)((0, semver_1.coerce)(version)), version } : null;
}
exports.getInstalledPackageVersionInfo = getInstalledPackageVersionInfo;
function addDependenciesToPackageJsonIfDontExist(tree, dependencies, devDependencies, packageJsonPath = 'package.json') {
    const packageJson = (0, devkit_1.readJson)(tree, packageJsonPath);
    function filterExisting(deps) {
        return Object.keys(deps)
            .filter((d) => { var _a, _b; return !((_a = packageJson.dependencies) === null || _a === void 0 ? void 0 : _a[d]) && !((_b = packageJson.devDependencies) === null || _b === void 0 ? void 0 : _b[d]); })
            .reduce((acc, d) => (Object.assign(Object.assign({}, acc), { [d]: deps[d] })), {});
    }
    const depsToAdd = filterExisting(dependencies);
    const devDepsToAdd = filterExisting(devDependencies);
    return (0, devkit_1.addDependenciesToPackageJson)(tree, depsToAdd, devDepsToAdd, packageJsonPath);
}
exports.addDependenciesToPackageJsonIfDontExist = addDependenciesToPackageJsonIfDontExist;
function versions(tree) {
    const majorAngularVersion = getInstalledAngularMajorVersion(tree);
    switch (majorAngularVersion) {
        case 14:
            return backward_compatible_versions_1.backwardCompatibleVersions.angularV14;
        case 15:
            return backward_compatible_versions_1.backwardCompatibleVersions.angularV15;
        default:
            return latestVersions;
    }
}
exports.versions = versions;
