// @nx/js/src/generators/library/library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.librarySchematic = exports.addLint = exports.projectGenerator = exports.libraryGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const ts_config_1 = require("../../utils/typescript/ts-config");
const path_1 = require("path");
const minimal_publish_script_1 = require("../../utils/minimal-publish-script");
const add_swc_config_1 = require("../../utils/swc/add-swc-config");
const add_swc_dependencies_1 = require("../../utils/swc/add-swc-dependencies");
const get_import_path_1 = require("../../utils/get-import-path");
const versions_1 = require("../../utils/versions");
const init_1 = require("../init/init");
const generator_1 = require("../setup-verdaccio/generator");
const create_ts_config_1 = require("../../utils/typescript/create-ts-config");
function libraryGenerator(tree, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { layoutDirectory, projectDirectory } = (0, devkit_1.extractLayoutDirectory)(schema.directory);
        schema.directory = projectDirectory;
        const libsDir = schema.rootProject
            ? '.'
            : layoutDirectory !== null && layoutDirectory !== void 0 ? layoutDirectory : (0, devkit_1.getWorkspaceLayout)(tree).libsDir;
        return projectGenerator(tree, schema, libsDir, (0, path_1.join)(__dirname, './files'));
    });
}
exports.libraryGenerator = libraryGenerator;
function projectGenerator(tree, schema, destinationDir, filesDir) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const tasks = [];
        tasks.push(yield (0, init_1.default)(tree, Object.assign(Object.assign({}, schema), { skipFormat: true, tsConfigName: schema.rootProject ? 'tsconfig.json' : 'tsconfig.base.json' })));
        const options = normalizeOptions(tree, schema, destinationDir);
        createFiles(tree, options, `${filesDir}/lib`);
        addProject(tree, options, destinationDir);
        tasks.push(addProjectDependencies(tree, options));
        if (options.publishable) {
            tasks.push(yield (0, generator_1.default)(tree, Object.assign(Object.assign({}, options), { skipFormat: true })));
        }
        if (options.bundler === 'vite') {
            const { viteConfigurationGenerator } = (0, devkit_1.ensurePackage)('@nx/vite', versions_1.nxVersion);
            const viteTask = yield viteConfigurationGenerator(tree, {
                project: options.name,
                newProject: true,
                uiFramework: 'none',
                includeVitest: options.unitTestRunner === 'vitest',
                includeLib: true,
                skipFormat: true,
            });
            tasks.push(viteTask);
        }
        if (options.bundler === 'rollup') {
            ensureBabelRootConfigExists(tree);
        }
        if (options.linter !== 'none') {
            const lintCallback = yield addLint(tree, options);
            tasks.push(lintCallback);
        }
        if (options.unitTestRunner === 'jest') {
            const jestCallback = yield addJest(tree, options);
            tasks.push(jestCallback);
            if (options.bundler === 'swc' || options.bundler === 'rollup') {
                replaceJestConfig(tree, options, `${filesDir}/jest-config`);
            }
        }
        else if (options.unitTestRunner === 'vitest' &&
            options.bundler !== 'vite' // Test would have been set up already
        ) {
            const { vitestGenerator } = (0, devkit_1.ensurePackage)('@nx/vite', versions_1.nxVersion);
            const vitestTask = yield vitestGenerator(tree, {
                project: options.name,
                uiFramework: 'none',
                coverageProvider: 'c8',
                skipFormat: true,
            });
            tasks.push(vitestTask);
        }
        if (!schema.skipTsConfig) {
            (0, ts_config_1.addTsConfigPath)(tree, options.importPath, [
                (0, devkit_1.joinPathFragments)(options.projectRoot, './src', 'index.' + (options.js ? 'js' : 'ts')),
            ]);
        }
        if (!options.skipFormat) {
            yield (0, devkit_1.formatFiles)(tree);
        }
        return (0, devkit_1.runTasksInSerial)(...tasks);
    });
}
exports.projectGenerator = projectGenerator;
function addProject(tree, options, destinationDir) {
    var _a;
    var _b;
    const projectConfiguration = {
        root: options.projectRoot,
        sourceRoot: (0, devkit_1.joinPathFragments)(options.projectRoot, 'src'),
        projectType: 'library',
        targets: {},
        tags: options.parsedTags,
    };
    if (options.bundler &&
        options.bundler !== 'none' &&
        options.config !== 'npm-scripts') {
        const outputPath = getOutputPath(options, destinationDir);
        projectConfiguration.targets.build = {
            executor: getBuildExecutor(options.bundler),
            outputs: ['{options.outputPath}'],
            options: {
                outputPath,
                main: `${options.projectRoot}/src/index` + (options.js ? '.js' : '.ts'),
                tsConfig: `${options.projectRoot}/tsconfig.lib.json`,
                assets: [],
            },
        };
        if (options.bundler === 'esbuild') {
            projectConfiguration.targets.build.options.generatePackageJson = true;
        }
        if (options.bundler === 'rollup') {
            projectConfiguration.targets.build.options.project = `${options.projectRoot}/package.json`;
            projectConfiguration.targets.build.options.compiler = 'swc';
        }
        if (options.bundler === 'swc' && options.skipTypeCheck) {
            projectConfiguration.targets.build.options.skipTypeCheck = true;
        }
        if (!options.minimal &&
            // TODO(jack): assets for rollup have validation that we need to fix (assets must be under <project-root>/src)
            options.bundler !== 'rollup') {
            (_a = (_b = projectConfiguration.targets.build.options).assets) !== null && _a !== void 0 ? _a : (_b.assets = []);
            projectConfiguration.targets.build.options.assets.push((0, devkit_1.joinPathFragments)(options.projectRoot, '*.md'));
        }
        if (options.publishable) {
            const publishScriptPath = (0, minimal_publish_script_1.addMinimalPublishScript)(tree);
            projectConfiguration.targets.publish = {
                command: `node ${publishScriptPath} ${options.name} {args.ver} {args.tag}`,
                dependsOn: ['build'],
            };
        }
    }
    if (options.config === 'workspace' || options.config === 'project') {
        (0, devkit_1.addProjectConfiguration)(tree, options.name, projectConfiguration);
    }
    else {
        (0, devkit_1.addProjectConfiguration)(tree, options.name, {
            root: projectConfiguration.root,
            tags: projectConfiguration.tags,
            targets: {},
        }, true);
    }
}
function addLint(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { lintProjectGenerator } = (0, devkit_1.ensurePackage)('@nx/linter', versions_1.nxVersion);
        return lintProjectGenerator(tree, {
            project: options.name,
            linter: options.linter,
            skipFormat: true,
            tsConfigPaths: [
                (0, devkit_1.joinPathFragments)(options.projectRoot, 'tsconfig.lib.json'),
            ],
            unitTestRunner: options.unitTestRunner,
            eslintFilePatterns: [
                `${options.projectRoot}/**/*.${options.js ? 'js' : 'ts'}`,
            ],
            setParserOptionsProject: options.setParserOptionsProject,
            rootProject: options.rootProject,
        });
    });
}
exports.addLint = addLint;
function updateTsConfig(tree, options) {
    (0, devkit_1.updateJson)(tree, (0, path_1.join)(options.projectRoot, 'tsconfig.json'), (json) => {
        if (options.strict) {
            json.compilerOptions = Object.assign(Object.assign({}, json.compilerOptions), { forceConsistentCasingInFileNames: true, strict: true, noImplicitOverride: true, noPropertyAccessFromIndexSignature: true, noImplicitReturns: true, noFallthroughCasesInSwitch: true });
        }
        return json;
    });
}
function addBabelRc(tree, options) {
    const filename = '.babelrc';
    const babelrc = {
        presets: [['@nx/js/babel', { useBuiltIns: 'usage' }]],
    };
    (0, devkit_1.writeJson)(tree, (0, path_1.join)(options.projectRoot, filename), babelrc);
}
function createFiles(tree, options, filesDir) {
    const { className, name, propertyName } = (0, devkit_1.names)(options.name);
    createProjectTsConfigJson(tree, options);
    (0, devkit_1.generateFiles)(tree, filesDir, options.projectRoot, Object.assign(Object.assign({}, options), { dot: '.', className,
        name,
        propertyName, js: !!options.js, cliCommand: 'nx', strict: undefined, tmpl: '', offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot), buildable: options.bundler && options.bundler !== 'none', hasUnitTestRunner: options.unitTestRunner !== 'none' }));
    if (options.bundler === 'swc' || options.bundler === 'rollup') {
        (0, add_swc_dependencies_1.addSwcDependencies)(tree);
        (0, add_swc_config_1.addSwcConfig)(tree, options.projectRoot, options.bundler === 'swc' ? 'commonjs' : 'es6');
    }
    else if (options.includeBabelRc) {
        addBabelRc(tree, options);
    }
    if (options.unitTestRunner === 'none') {
        tree.delete((0, path_1.join)(options.projectRoot, 'src/lib', `${options.fileName}.spec.ts`));
        tree.delete((0, path_1.join)(options.projectRoot, 'src/app', `${options.fileName}.spec.ts`));
    }
    if (options.js) {
        (0, devkit_1.toJS)(tree);
    }
    const packageJsonPath = (0, devkit_1.joinPathFragments)(options.projectRoot, 'package.json');
    if (tree.exists(packageJsonPath)) {
        (0, devkit_1.updateJson)(tree, packageJsonPath, (json) => {
            json.name = options.importPath;
            json.version = '0.0.1';
            json.type = 'commonjs';
            // If the package is publishable, we should remove the private field.
            if (json.private && options.publishable) {
                delete json.private;
            }
            return json;
        });
    }
    else {
        (0, devkit_1.writeJson)(tree, packageJsonPath, {
            name: options.importPath,
            version: '0.0.1',
            type: 'commonjs',
        });
    }
    if (options.config === 'npm-scripts') {
        (0, devkit_1.updateJson)(tree, packageJsonPath, (json) => {
            json.scripts = {
                build: "echo 'implement build'",
                test: "echo 'implement test'",
            };
            return json;
        });
    }
    else if ((!options.bundler || options.bundler === 'none') &&
        !(options.projectRoot === '.')) {
        tree.delete(packageJsonPath);
    }
    if (options.minimal && !(options.projectRoot === '.')) {
        tree.delete((0, path_1.join)(options.projectRoot, 'README.md'));
    }
    updateTsConfig(tree, options);
}
function addJest(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { jestProjectGenerator } = (0, devkit_1.ensurePackage)('@nx/jest', versions_1.nxVersion);
        return yield jestProjectGenerator(tree, Object.assign(Object.assign({}, options), { project: options.name, setupFile: 'none', supportTsx: false, skipSerializers: true, testEnvironment: options.testEnvironment, skipFormat: true, compiler: options.bundler === 'swc' || options.bundler === 'tsc'
                ? options.bundler
                : options.bundler === 'rollup'
                    ? 'swc'
                    : undefined }));
    });
}
function replaceJestConfig(tree, options, filesDir) {
    // the existing config has to be deleted otherwise the new config won't overwrite it
    const existingJestConfig = (0, devkit_1.joinPathFragments)(filesDir, `jest.config.${options.js ? 'js' : 'ts'}`);
    if (tree.exists(existingJestConfig)) {
        tree.delete(existingJestConfig);
    }
    // replace with JS:SWC specific jest config
    (0, devkit_1.generateFiles)(tree, filesDir, options.projectRoot, {
        ext: options.js ? 'js' : 'ts',
        js: !!options.js,
        project: options.name,
        offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot),
        projectRoot: options.projectRoot,
        testEnvironment: options.testEnvironment,
    });
}
function normalizeOptions(tree, options, destinationDir) {
    /**
     * We are deprecating the compiler and the buildable options.
     * However, we want to keep the existing behavior for now.
     *
     * So, if the user has not provided a bundler, we will use the compiler option, if any.
     *
     * If the user has not provided a bundler and no compiler, but has set buildable to true,
     * we will use tsc, since that is the compiler the old generator used to default to, if buildable was true
     * and no compiler was provided.
     *
     * If the user has not provided a bundler and no compiler, and has not set buildable to true, then
     * set the bundler to tsc, to preserve old default behaviour (buildable: true by default).
     *
     * If it's publishable, we need to build the code before publishing it, so again
     * we default to `tsc`. In the previous version of this, it would set `buildable` to true
     * and that would default to `tsc`.
     *
     * In the past, the only way to get a non-buildable library was to set buildable to false.
     * Now, the only way to get a non-buildble library is to set bundler to none.
     * By default, with nothing provided, libraries are buildable with `@nx/js:tsc`.
     */
    var _a, _b, _c;
    options.bundler = (_b = (_a = options.bundler) !== null && _a !== void 0 ? _a : options.compiler) !== null && _b !== void 0 ? _b : 'tsc';
    // ensure programmatic runs have an expected default
    if (!options.config) {
        options.config = 'project';
    }
    if (options.publishable) {
        if (!options.importPath) {
            throw new Error(`For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`);
        }
        if (options.bundler === 'none') {
            options.bundler = 'tsc';
        }
    }
    // This is to preserve old behaviour, buildable: false
    if (options.publishable === false && options.buildable === false) {
        options.bundler = 'none';
    }
    const { Linter } = (0, devkit_1.ensurePackage)('@nx/linter', versions_1.nxVersion);
    if (options.config === 'npm-scripts') {
        options.unitTestRunner = 'none';
        options.linter = Linter.None;
        options.bundler = 'none';
    }
    if ((options.bundler === 'swc' || options.bundler === 'rollup') &&
        options.skipTypeCheck == null) {
        options.skipTypeCheck = false;
    }
    const name = (0, devkit_1.names)(options.name).fileName;
    const projectDirectory = options.directory
        ? `${(0, devkit_1.names)(options.directory).fileName}/${name}`
        : options.rootProject
            ? '.'
            : name;
    if (!options.unitTestRunner && options.bundler === 'vite') {
        options.unitTestRunner = 'vitest';
    }
    else if (!options.unitTestRunner && options.config !== 'npm-scripts') {
        options.unitTestRunner = 'jest';
    }
    if (!options.linter && options.config !== 'npm-scripts') {
        options.linter = Linter.EsLint;
    }
    const projectName = options.rootProject
        ? name
        : projectDirectory.replace(new RegExp('/', 'g'), '-');
    const fileName = getCaseAwareFileName({
        fileName: options.simpleName ? name : projectName,
        pascalCaseFiles: options.pascalCaseFiles,
    });
    const projectRoot = (0, devkit_1.joinPathFragments)(destinationDir, projectDirectory);
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    const importPath = options.importPath || (0, get_import_path_1.getImportPath)(tree, projectDirectory);
    (_c = options.minimal) !== null && _c !== void 0 ? _c : (options.minimal = false);
    return Object.assign(Object.assign({}, options), { fileName, name: projectName, projectRoot,
        projectDirectory,
        parsedTags,
        importPath });
}
function getCaseAwareFileName(options) {
    const normalized = (0, devkit_1.names)(options.fileName);
    return options.pascalCaseFiles ? normalized.className : normalized.fileName;
}
function addProjectDependencies(tree, options) {
    if (options.bundler == 'esbuild') {
        return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, {
            '@nx/esbuild': versions_1.nxVersion,
            '@types/node': versions_1.typesNodeVersion,
            esbuild: versions_1.esbuildVersion,
        });
    }
    else if (options.bundler == 'rollup') {
        return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, { '@nx/rollup': versions_1.nxVersion, '@types/node': versions_1.typesNodeVersion });
    }
    else {
        return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, { '@types/node': versions_1.typesNodeVersion });
    }
    // Vite is being installed in the next step if bundler is vite
    // noop
    return () => { };
}
function getBuildExecutor(bundler) {
    switch (bundler) {
        case 'esbuild':
            return `@nx/esbuild:esbuild`;
        case 'rollup':
            return `@nx/rollup:rollup`;
        case 'swc':
        case 'tsc':
            return `@nx/js:${bundler}`;
        case 'vite':
            return `@nx/vite:build`;
        case 'none':
        default:
            return undefined;
    }
}
function ensureBabelRootConfigExists(tree) {
    if (tree.exists('babel.config.json'))
        return;
    (0, devkit_1.writeJson)(tree, 'babel.config.json', {
        babelrcRoots: ['*'],
    });
}
function getOutputPath(options, destinationDir) {
    const parts = ['dist'];
    if (destinationDir) {
        parts.push(destinationDir);
    }
    if (options.projectDirectory === '.') {
        parts.push(options.name);
    }
    else {
        parts.push(options.projectDirectory);
    }
    return (0, devkit_1.joinPathFragments)(...parts);
}
function createProjectTsConfigJson(tree, options) {
    const tsconfig = {
        extends: options.rootProject
            ? undefined
            : (0, ts_config_1.getRelativePathToRootTsConfig)(tree, options.projectRoot),
        compilerOptions: Object.assign(Object.assign({}, (options.rootProject ? create_ts_config_1.tsConfigBaseOptions : {})), { module: 'commonjs', allowJs: options.js ? true : undefined }),
        files: [],
        include: [],
        references: [
            {
                path: './tsconfig.lib.json',
            },
        ],
    };
    (0, devkit_1.writeJson)(tree, (0, devkit_1.joinPathFragments)(options.projectRoot, 'tsconfig.json'), tsconfig);
}
exports.default = libraryGenerator;
exports.librarySchematic = (0, devkit_1.convertNxGenerator)(libraryGenerator);
