// @nx/angular/src/generators/setup-mf/lib/setup-serve-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupServeTarget = void 0;
const devkit_1 = require("@nx/devkit");
function setupServeTarget(host, options) {
    var _a, _b;
    const appConfig = (0, devkit_1.readProjectConfiguration)(host, options.appName);
    appConfig.targets['serve'] = Object.assign(Object.assign({}, appConfig.targets['serve']), { executor: options.mfType === 'host'
            ? '@nx/angular:module-federation-dev-server'
            : '@nx/angular:webpack-dev-server', options: Object.assign(Object.assign({}, appConfig.targets['serve'].options), { port: (_a = options.port) !== null && _a !== void 0 ? _a : undefined, publicHost: `http://localhost:${(_b = options.port) !== null && _b !== void 0 ? _b : 4200}` }) });
    if (options.mfType === 'remote') {
        appConfig.targets['serve-static'] = {
            executor: '@nx/web:file-server',
            defaultConfiguration: 'development',
            options: {
                buildTarget: `${options.appName}:build`,
                port: options.port,
                watch: false,
            },
            configurations: {
                development: {
                    buildTarget: `${options.appName}:build:development`,
                },
                production: {
                    buildTarget: `${options.appName}:build:production`,
                },
            },
        };
    }
    (0, devkit_1.updateProjectConfiguration)(host, options.appName, appConfig);
}
exports.setupServeTarget = setupServeTarget;
